autoPyTorch/data/tabular_feature_validator.py

```diff
@@ -337,16 +337,14 @@ def _check_data(
 
             dtypes = [dtype.name for dtype in X.dtypes]
 
-            dtypes_diff = [s_dtype != dtype for s_dtype, dtype in zip(self.dtypes, dtypes)]
+            diff_cols = X.columns[[s_dtype != dtype for s_dtype, dtype in zip(self.dtypes, dtypes)]]
             if len(self.dtypes) == 0:
                 self.dtypes = dtypes
-            elif (
-                any(dtypes_diff)  # the dtypes of some columns are different in train and test dataset
-                and self.all_nan_columns is not None  # Ignore all_nan_columns is None
-                and len(set(X.columns[dtypes_diff]).difference(self.all_nan_columns)) != 0
-            ):
-                # The dtypes can be different if and only if the column belongs
-                # to all_nan_columns as these columns would be imputed.
+            elif not self._is_datasets_consistent(diff_cols, X):
+                # The dtypes can be different if the column belongs
+                # to all_nan_columns if the dataset is in training
+                # set or if the column is all nan if the dataset
+                # is for testing as these columns would be imputed.
```

Comment:
Probably, we can delete those comments or move them to the doc-string of `is_datasets_consistent`?

Reply:
yes I was thinking the same thing. But I wanted to wait for the tests to pass. I'll change this if they pass now.